core/src/packet_forward_manager.rs

```diff
@@ -0,0 +1,427 @@
+                .into_iter()
+                .map(|_| Vec::<&'a Packet>::new())
+                .collect(),
+        }
+    }
+
+    pub fn take(
+        &mut self,
+        unprocessed_packet_batches: &'a UnprocessedPacketBatches,
+        working_bank: Arc<Bank>,
+        qos_service: &'a QosService,
+    ) -> Vec<&'a Packet> {
+        let prioritized_forwardable_packet_locators = Self::prioritize_unforwarded_packets_by_fee(
+            unprocessed_packet_batches,
+            Some(working_bank.clone()),
+        );
+
+        // if we have a bank to work with, then sort packets by write-account buckets
+        let (transactions, sanitized_locators) = sanitize_transactions(
+            unprocessed_packet_batches,
+            &prioritized_forwardable_packet_locators,
+            &working_bank.feature_set,
+            working_bank.vote_only_bank(),
+            working_bank.as_ref(),
+        );
+        let transactions_costs = qos_service.compute_transaction_costs(transactions.iter());
+
+        transactions
+            .iter()
+            .zip(transactions_costs.iter())
+            .zip(sanitized_locators.iter())
+            .for_each(|((tx, cost), packet_locator)| {
+                if let Some(packet) =
+                    Self::locate_packet(unprocessed_packet_batches, packet_locator)
+                {
+                    self.sort_into_buckets(tx, cost, packet)
+                }
+            });
+
+        let mut result = Vec::<&'a Packet>::new();
+        self.forwardable_packets
+            .iter()
+            .for_each(|v| result.extend(v));
+        result
+    }
+
+    // prioritize unforwarded, unprocessed packets in buffered packet_batches by its fee/CU
+    fn prioritize_unforwarded_packets_by_fee(
+        unprocessed_packet_batches: &'a UnprocessedPacketBatches,
+        working_bank: Option<Arc<Bank>>,
+    ) -> Vec<PacketLocator> {
+        let mut locators = Vec::<PacketLocator>::with_capacity(TOTAL_BUFFERED_PACKETS);
+        unprocessed_packet_batches.iter().enumerate().for_each(
+            |(batch_index, deserialized_packet_batch)| {
+                if !deserialized_packet_batch.forwarded {
+                    deserialized_packet_batch
+                        .unprocessed_packets
+                        .keys()
+                        .for_each(|packet_index| {
+                            locators.push(PacketLocator {
+                                batch_index,
+                                packet_index: *packet_index,
+                            });
+                        })
+                }
+            },
+        );
+
+        prioritize_by_fee(unprocessed_packet_batches, &locators, working_bank)
+    }
+
+    fn sort_into_buckets(
+        &mut self,
+        transaction: &SanitizedTransaction,
+        cost: &TransactionCost,
+        packet: &'a Packet,
+    ) {
+        // try to sort the `transaction` into one of outbound (virtual) blocks
```

Comment:
It's unclear what a "virtual" block is here, a more concrete definition would be helpful here